plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java

```diff
@@ -0,0 +1,235 @@
+        maxQueryGroupCount = MAX_QUERY_GROUP_COUNT.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(MAX_QUERY_GROUP_COUNT, this::setMaxQueryGroupCount);
+        inflightCreateQueryGroupRequestCount = new AtomicInteger();
+        inflightResourceLimitValues = new HashMap<>();
+    }
+
+    /**
+     * Set maxQueryGroupCount to be newMaxQueryGroupCount
+     * @param newMaxQueryGroupCount - the max number of QueryGroup allowed
+     */
+    public void setMaxQueryGroupCount(int newMaxQueryGroupCount) {
+        if (newMaxQueryGroupCount < 0) {
+            throw new IllegalArgumentException("node.query_group.max_count can't be negative");
+        }
+        this.maxQueryGroupCount = newMaxQueryGroupCount;
+    }
+
+    @Override
+    public void persist(QueryGroup queryGroup, ActionListener<CreateQueryGroupResponse> listener) {
+        persistInClusterStateMetadata(queryGroup, listener);
+    }
+
+    /**
+     * Update cluster state to include the new QueryGroup
+     * @param queryGroup {@link QueryGroup} - the QueryGroup we're currently creating
+     */
+    void persistInClusterStateMetadata(QueryGroup queryGroup, ActionListener<CreateQueryGroupResponse> listener) {
+        clusterService.submitStateUpdateTask(SOURCE, new ClusterStateUpdateTask(Priority.URGENT) {
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                return saveQueryGroupInClusterState(queryGroup, currentState);
+            }
+
+            @Override
+            public ThrottlingKey getClusterManagerThrottlingKey() {
+                return createQueryGroupThrottlingKey;
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                restoreInflightValues(queryGroup.getResourceLimits());
+                inflightCreateQueryGroupRequestCount.decrementAndGet();
+                logger.warn("failed to save QueryGroup object due to error: {}, for source: {}", e.getMessage(), source);
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                restoreInflightValues(queryGroup.getResourceLimits());
+                inflightCreateQueryGroupRequestCount.decrementAndGet();
+                CreateQueryGroupResponse response = new CreateQueryGroupResponse(queryGroup, RestStatus.OK);
+                listener.onResponse(response);
+            }
+        });
+    }
+
+    /**
+     * Get the allocation value for resourceName for the QueryGroup
+     * @param resourceName - the resourceName we want to get the usage for
+     * @param resourceLimits  - the resource limit from which to get the allocation value for resourceName
+     */
+    private double getResourceLimitValue(String resourceName, final Map<ResourceType, Object> resourceLimits) {
```

Comment:
I think it stayed because initially we were using `XContentParser.map()` to parse the resource limits object. I think we can keep it as `Map<String, Double>` or `Map<String, ResourceValue>` to do resource specific parsing